### PR TITLE
Feature/lixin/gp from aimd

### PR DESCRIPTION
**Pull request by nw13slx**
Slight modification and clean up for gp_from_aimd module and output module

- min_atoms_added function enable. training is only called when added atoms are more than this parameter
- hyper parameters printing is more formatted in gp_from_aimd and gp_algrebra
- output.py is cleaned up. replacing all the string format into f-string.

### CODE REVIEW

File: flare/gp_algebra.py [hunk 1]
```diff
@@ -279,7 +276,7 @@ def get_neg_likelihood(hyps: np.ndarray, training_data: list,
     like = get_like_from_ky_mat(ky_mat, training_labels_np)
 
     if output is not None:
-        output.write_to_log('like: ' + str(like)+'\n', name="hyps")
+        output.write_hyps(None, hyps, None, like, "NA", name="hyps")
```
Comment (stevetorr) [inline flare/gp_algebra.py]: Achtung! write_hyps has no "name" kwarg:
https://github.com/mir-group/flare/blob/master/flare/output.py#L264
:)

File: flare/gp_algebra.py [hunk 2]
```diff
@@ -305,7 +299,6 @@ def get_neg_like_grad(hyps: np.ndarray, training_data: list,
         get_like_grad_from_mats(ky_mat, hyp_mat, training_labels_np)
 
     if output is not None:
-        output.write_to_log('like grad: ' + str(like_grad)+'\n', name="hyps")
-        output.write_to_log('like: ' + str(like)+'\n', name="hyps")
+        output.write_hyps(None, hyps, None, like, like_grad, name="hyps")
```
Comment (stevetorr) [inline flare/gp_algebra.py]: No hyperparameter labels passed here?

File: flare/gp_from_aimd.py [hunk 3]
```diff
@@ -4,18 +4,20 @@
...
+import numpy as np
+
 from flare.output import Output
+from flare.struc import Structure, get_unique_species
+from flare.gp import GaussianProcess
+from flare.env import AtomicEnvironment
+from flare.util import Z_to_element
+from flare.predict import predict_on_structure, \
+    predict_on_structure_par, predict_on_structure_en, \
+    predict_on_structure_par_en
```
Comment (stevetorr) [inline flare/gp_from_aimd.py]: Thanks for cleaning this up!

File: flare/gp_from_aimd.py [hunk 4]
```diff
@@ -189,14 +195,30 @@ def run(self):
...
+                count_perspecies[ele] = 0
+            print(list(mae_perspecies.keys()))
+            for atom in range(cur_frame.nat):
+                Z = cur_frame.coded_species[atom]
+                ele = Z_to_element(Z)
+                mae_perspecies[ele] += np.sum(error[atom, :])
+                count_perspecies[ele] += 1
+            for ele in species:
+                mae_perspecies[ele] /= (count_perspecies[ele]*3+1e-25)
+
```
Comment (stevetorr) [inline flare/gp_from_aimd.py]: Thanks for implementing this! This will be a massive boon to myself and other users.

File: flare/gp_from_aimd.py [hunk 5]
```diff
@@ -189,14 +195,30 @@ def run(self):
...
-            mae = np.mean(np.abs(cur_frame.forces - dft_forces)) * 1000
+            error = np.abs(cur_frame.forces - dft_forces)
+            mae = np.mean(error) * 1000
+            mae_perspecies = {}
+            count_perspecies = {}
+            species = [Z_to_element(Z) for Z in set(cur_frame.coded_species)]
+            for ele in species:
+                mae_perspecies[ele] = 0
+                count_perspecies[ele] = 0
+            print(list(mae_perspecies.keys()))
```
Comment (stevetorr) [inline flare/gp_from_aimd.py]: Was this a debug thing that got left in by mistake? If not, should add an if verbose: flag.

File: flare/output.py [hunk 6]
```diff
@@ -271,27 +265,33 @@ def write_hyps(self, hyp_labels, hyps, start_time, like, like_grad):
...
+            time_curr = time.time() - start_time
+            f.write('wall time from start: {time_curr:.2} s \n')
 
         if self.always_flush:
             f.flush()
 
     def write_gp_dft_comparison(self, curr_step, frame,
                                 start_time, dft_forces,
-                                mae, mac, local_energies=None, KE=None):
+                                mae, pmae, mac, local_energies=None, KE=None):
```
Comment (stevetorr) [inline flare/output.py]: Would suggest renaming pmae to mae_ps to mirror 'mae_perspecies"

**Comment by stevetorr** [comment]
Looks great!

File: flare/gp_algebra.py (hunk 2 above)
Comment (stevetorr) [reply flare/gp_algebra.py]: I guess that's the way we've been doing it but it might be useful.

Comment (nw13slx) [reply flare/gp_algebra.py]: the label is not available in the function and I didn't want to change the argument for the function

File: flare/gp_from_aimd.py (hunk 5 above)
Comment (nw13slx) [reply flare/gp_from_aimd.py]: oh yeah. forgot to remove this line...

File: flare/output.py (hunk 6 above)
Comment (nw13slx) [reply flare/output.py]: fixed in efd157402d391ec9091bb5041bd8a6bddf753c22

File: flare/gp_from_aimd.py (hunk 5 above)
Comment (nw13slx) [reply flare/gp_from_aimd.py]: fixed in efd157402d391ec9091bb5041bd8a6bddf753c22

**Pull request merged**